File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution { 
    TreeNode prev; 
    public boolean isValidBST(TreeNode root) { 
        prev = null;
        return inorder(root); 
    } 

    public boolean inorder(TreeNode root){ 
        if(root == null){
            return true;
        }

        if(inorder(root.left) == false){
            return false;
        }
        
        if(prev != null && prev.val >= root.val){
            return false;
        }
        
        prev = root;

        return (inorder(root.right));
    }
}

File: 0098-validate-binary-search-tree/0098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def __init__(self):
        self.prev = None

    def isValidBST(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: bool
        """
        return self.inorder(root)

    def inorder(self, root):
        if not root:
            return True

        if not self.inorder(root.left):  
            return False

        if self.prev and self.prev.val >= root.val:
            return False

        self.prev = root

        return self.inorder(root.right)
